### PR TITLE
Updating validation for form

### PR DESCRIPTION
**Pull request by brett-onions**
## Summary by CodeRabbit

- **New Features**
	- Enhanced validation logic for role name input, providing immediate feedback on validity.
	- Introduced a new validation schema for user data using the `zod` library.
	- Added functions for structured state updates and robust validation mechanisms for user forms.

- **Bug Fixes**
	- Improved error handling for input fields, ensuring users cannot submit empty role names.

- **Documentation**
	- Updated README to clarify the command for starting the development instance and concurrency settings.

- **Chores**
	- Added new dependencies and updated TypeScript configuration for modern JavaScript features.

### CODE REVIEW

File: packages/users-app/src/screens/helper.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import { SelectChangeEvent } from "@mui/material"
+import { User } from '../types'
+import { userSchema } from "../utils"
+
+export function handleOnChange(nestedKey: string, user: Readonly<User>, e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement> | SelectChangeEvent<string[]>, setUser: React.Dispatch<React.SetStateAction<Readonly<import("/home/brett/GitHub/openhim-console/packages/users-app/src/types/index").User>>>, validateUserField: (field: string, newBasicInfoState?: object) => void) {
```
Comment (marvinIsSacul) [inline packages/users-app/src/screens/helper.ts:5]: what the GitHub import for?

**Review by marvinIsSacul — COMMENTED** [review]
couple of comments.

**Review by ItsMurumba — APPROVED** [review]
LGTM!

File: packages/rbac-app/src/screens/steps/channels.and.clients.step.tsx [hunk 2]
```diff
@@ -59,6 +60,15 @@ export function ChannelsClientsStep(props: {
 
   const paperStyling = {padding: '10px 20px'}
 
+  const validateClientName = (newState: Role) => {
+    console.log(newState.name.trim());
```
Comment (marvinIsSacul) [inline packages/rbac-app/src/screens/steps/channels.and.clients.step.tsx:64]:
```suggestion
```

File: packages/users-app/src/components/common/basic.info.component.tsx [hunk 3]
```diff
@@ -64,23 +61,20 @@ export function BasicInfo(props: {
...
-            helperText={
-              user.firstname.trim() === ''
-                ? 'First Name cannot be empty'
-                : undefined
-            }
+            name="firstname"
+            value={props.user.firstname}
+            onChange={props.onChange}
+            onBlur={e => props.validateUserField('firstname')}
+            error={props.validationErrors.firstname ? true : false}
```
Comment (marvinIsSacul) [inline packages/users-app/src/components/common/basic.info.component.tsx:76]:
```suggestion
            error={!!props.validationErrors.firstname}
```

File: packages/users-app/src/screens/helper.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import { User } from '../types'
+import { userSchema } from "../utils"
+
+export function handleOnChange(nestedKey: string, user: Readonly<User>, e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement> | SelectChangeEvent<string[]>, setUser: React.Dispatch<React.SetStateAction<Readonly<User>>>, validateUserField: (field: string, newBasicInfoState?: object) => void) {
+    let newBasicUserState
+  
+    if (nestedKey) {
+      newBasicUserState = {
+        ...user
+      }
```
Comment (marvinIsSacul) [inline packages/users-app/src/screens/helper.ts:11]:
```suggestion
      newBasicUserState = structuredClone(user)
```

Safer deep cloning. It's helped me resolve quite a lot of bugs.

**Review by marvinIsSacul — APPROVED** [review]
Looks okay. Just a couple of suggestions.

File: packages/users-app/src/screens/helper.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      const keys = nestedKey.split('.')
+      newBasicUserState[keys[0]][keys[1]][keys[2]] =
+        !newBasicUserState[keys[0]][keys[1]][keys[2]]
+    } else {
+      const field = e.target.name
+      if (field === 'dailyReport' || field === 'weeklyReport') {
+        console.log('field', field)
```
Comment (marvinIsSacul) [inline packages/users-app/src/screens/helper.ts:18]:
```suggestion
```

File: packages/users-app/src/utils/index.ts [hunk 6]
```diff
@@ -1,4 +1,34 @@
...
+  }),
+  token: z.string().optional(),
+  tokenType: z.string().nullable().optional()
+})
+
+
+
 
 export const defaultUser: Readonly<User> = {
   _id: '',
```
Comment (marvinIsSacul) [inline packages/users-app/src/utils/index.ts:34]:
```suggestion
  _id: undefined,
`

**Review by drizzentic — APPROVED** [review]
LGTM!

**Comment by drizzentic** [comment]
@coderabbitai review

**Review by ItsMurumba — APPROVED** [review]
LGTM

**Pull request merged**